File: packages/shared-types/src/payments.ts
import type { BookingDraftSummary, DepositStatus } from "./bookings";
import type { ActorSummary, AuditFields, ISODateString, TenantScoped, UUID } from "./common";

export type PaymentStatus =
  | "pending"
  | "requires_action"
  | "processing"
  | "succeeded"
  | "failed"
  | "refunded"
  | "canceled";

export type PaymentMethodType = "card" | "cash" | "external_pos" | "wallet_credit" | "manual";

export type CheckoutSessionKind = "deposit" | "booking_balance" | "manual_payment_link";
export type DepositPaymentLinkState = "open" | "expired" | "missing";

export type PaymentEventKind =
  | "checkout_started"
  | "checkout_completed"
  | "checkout_expired"
  | "checkout_reminder_sent"
  | "payment_recorded"
  | "refund_recorded"
  | "deposit_forfeited"
  | "manual_credit_applied";

export type PaymentEvent = {
  id: UUID;
  kind: PaymentEventKind;
  occurredAt: ISODateString;
  actor: ActorSummary;
  amountCents?: number;
  notes?: string;
  stripeSessionId?: string;
  stripePaymentIntentId?: string;
};

export type PaymentSummary = AuditFields &
  TenantScoped & {
    bookingId?: UUID | null;
    bookingDraftId?: UUID | null;
    customerId: UUID;
    status: PaymentStatus;
    depositStatus: DepositStatus;
    amountCents: number;
    currency: string;
    paymentMethodType: PaymentMethodType;
    checkoutSessionKind?: CheckoutSessionKind | null;
    stripeSessionId?: string | null;
    stripePaymentIntentId?: string | null;
    latestEvent?: PaymentEvent | null;
    events: PaymentEvent[];
  };

export type CreateCheckoutSessionRequest = {
  tenantSlug: string;
  bookingDraftId?: UUID;
  bookingId?: UUID;
  kind: CheckoutSessionKind;
  successUrl: string;
  cancelUrl: string;
};

export type CreateCheckoutSessionResponse = {
  checkoutUrl: string;
  sessionId: string;
  expiresAt?: ISODateString;
};

export type DepositPaymentFollowUpItem = {
  bookingDraft: BookingDraftSummary;
  paymentId?: UUID | null;
  paymentStatus?: PaymentStatus | null;
  checkoutSessionId?: string | null;
  checkoutUrl?: string | null;
  checkoutExpiresAt?: ISODateString | null;
  linkState: DepositPaymentLinkState;
};

export type DepositPaymentFollowUpListResponse = {
  items: DepositPaymentFollowUpItem[];
};

export type SendPaymentReminderResponse = {
  bookingDraftId: UUID;
  paymentId: UUID;
  checkoutSessionId: string;
  checkoutUrl: string;
  recipientEmail: string;
  provider: string;
  providerMessageId: string;
  sentAt: ISODateString;
};

export type RecordManualPaymentRequest = {
  amountCents: number;
  paymentMethodType: string;
  notes?: string;
};

File: backend/app/schemas/payments.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.base import CamelModel
from app.schemas.booking_drafts import BookingDraftSummaryResponse


CheckoutSessionKind = Literal["deposit", "booking_balance", "manual_payment_link"]
DepositPaymentLinkState = Literal["open", "expired", "missing"]


class CreateCheckoutSessionRequest(CamelModel):
    tenant_slug: str
    booking_draft_id: str | None = None
    booking_id: str | None = None
    kind: CheckoutSessionKind
    success_url: str
    cancel_url: str


class CreateCheckoutSessionResponse(CamelModel):
    checkout_url: str
    session_id: str
    expires_at: datetime | None = None


class DepositPaymentFollowUpItemResponse(CamelModel):
    booking_draft: BookingDraftSummaryResponse
    payment_id: str | None = None
    payment_status: str | None = None
    checkout_session_id: str | None = None
    checkout_url: str | None = None
    checkout_expires_at: datetime | None = None
    link_state: DepositPaymentLinkState


class DepositPaymentFollowUpListResponse(CamelModel):
    items: list[DepositPaymentFollowUpItemResponse]


class SendPaymentReminderResponse(CamelModel):
    booking_draft_id: str
    payment_id: str
    checkout_session_id: str
    checkout_url: str
    recipient_email: str
    provider: str
    provider_message_id: str
    sent_at: datetime


class RecordManualPaymentRequest(CamelModel):
    amount_cents: int = Field(gt=0)
    payment_method_type: str
    notes: str | None = Field(default=None, max_length=500)


class ApplyWalletCreditRequest(CamelModel):
    amount_cents: int = Field(gt=0)